### PR TITLE
Keygen

### PR DESCRIPTION
**Pull request by Hey**

### CODE REVIEW

File: internal/types/vault.go [hunk 1]
```diff
@@ -25,11 +23,11 @@ type VaultCreateResponse struct {
 }
 
 func (v *VaultCreateResponse) Task() (*asynq.Task, error) {
-	payload, err := json.Marshal(v)
+	task, err := tasks.NewKeyGeneration("Vultisigner", v.SessionID, v.HexChainCode)
```
Comment (johnnyluo) [inline internal/types/vault.go:26]: I think you also need `HexEncryptionKey` in this , otherwise how the worker decrypt the message , and apply to TSS instance?

File: storage/redis_storage.go [hunk 2]
```diff
@@ -66,7 +66,8 @@ func (r *RedisStorage) GetUser(ctx context.Context, name string) (string, error)
 	if contexthelper.CheckCancellation(ctx) != nil {
 		return "", ctx.Err()
 	}
-	return r.client.Get(ctx, name).Result()
+	result, error := r.client.Get(ctx, name).Result()
```
Comment (johnnyluo) [inline storage/redis_storage.go:69]: why this change?

File: api/server.go [hunk 3]
```diff
@@ -57,7 +57,7 @@ func (s *Server) StartServer() error {
 	grp.POST("/create", s.CreateVault)
 	grp.POST("/upload", s.UploadVault)
 	grp.GET("/download/{publicKeyECDSA}", s.DownloadVault)
-	return e.Start(fmt.Sprintf(":%d", s.port))
+	return e.Start(fmt.Sprintf("localhost:%d", s.port))
```
Comment (johnnyluo) [inline api/server.go:60]: if you specify `localhost` then the port will only open on your loopback network interface,  it won't be able to accept traffic from outside.

File: internal/keygen/keygen.go [hunk 4]
```diff
@@ -1,25 +1,91 @@
...
+	// 	Session:   kg.Session,
+	// 	ChainCode: kg.ChainCode,
+	// })
+	// if err != nil {
+	// 	return err
+	// }
+
+	// return nil
+
+	if err := coordinator.RegisterSession(relayServer, kg.Session, kg.Key); err != nil {
```
Comment (johnnyluo) [inline internal/keygen/keygen.go:34]: make coordinator part of vultisigner ,  don't rely on mobile-tss-lib to provide coordinate feature

File: go.mod [hunk 5]
```diff
@@ -75,4 +75,6 @@ require (
 replace (
 	github.com/agl/ed25519 => github.com/binance-chain/edwards25519 v0.0.0-20200305024217-f36fc4b53d43
 	github.com/gogo/protobuf => github.com/gogo/protobuf v1.3.2
+	// DEVELOPMENT
+	github.com/vultisig/mobile-tss-lib => ../mobile-tss-lib
```
Comment (johnnyluo) [inline go.mod:79]: remove this

File: relay/messenger.go [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+					log.Println("fail to decode messages:", err)
+				}
+				continue
+			}
+			for _, message := range messages {
+				if message.From == key {
+					continue
+				}
+
+				hash := md5.Sum([]byte(message.Body))
```
Comment (johnnyluo) [inline relay/messenger.go:110]: You don't need to calculate the hash here again , there is field call `hash` in the message's json payload, just use that one

File: relay/messenger.go [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+				hashStr := hex.EncodeToString(hash[:])
+
+				client := http.Client{}
+				req, err := http.NewRequest(http.MethodDelete, server+"/message/"+session+"/"+key+"/"+hashStr, nil)
```
Comment (johnnyluo) [inline relay/messenger.go:114]: You should always apply it to local instance before deleting it.

File: relay/messenger.go (hunk 6 above)
Comment (johnnyluo) [inline relay/messenger.go:110]: the message is encrypted using the hexecnryptionkey , you need to decrypted first

Comment (johnnyluo) [inline relay/messenger.go:110]: You also need to implement a in memory cache to keep track the message that has been applied before , see the implementation in IOS or android https://github.com/vultisig/vultisig-android/blob/main/app/src/main/java/com/vultisig/wallet/tss/TssMessagePuller.kt#L62

File: relay/session.go [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+	if resp.StatusCode != http.StatusCreated {
+		return fmt.Errorf("fail to start session: %s", resp.Status)
+	}
+	// session registered , and device can use this session id to join the session
+	return nil
+}
+
+func (s *Server) RegisterSession(sessionID string, key string) error {
+	sessionURL := s.vultisigRelay + "/" + sessionID
+	fmt.Println("Registering session with url: ", sessionURL)
```
Comment (johnnyluo) [inline relay/session.go:39]: This is not production ready code , `fmt.Println` everywhere , if you need to log those messages , use the logging library we have in place, otherwise  remove it

**Review by johnnyluo — CHANGES_REQUESTED** [review]

File: relay/messenger.go (hunk 6 above)
Comment (johnnyluo) [reply relay/messenger.go:110]: the hash is calculated based on an already encrypted body , I don't think it will or matter, if someone tampered it , then the description will fail

File: relay/state.go [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+	}
+	return nil
+}
+
+func (l *LocalStateAccessorImp) GetLocalState(pubKey string) (string, error) {
+	if err := l.ensureFolder(); err != nil {
+		return "", err
+	}
+	fileName := filepath.Join(l.Folder, pubKey+"-"+l.Key+".json")
+	fmt.Println(fileName)
```
Comment (johnnyluo) [inline relay/state.go:30]: remove

File: relay/state.go [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+	if _, err := os.Stat(fileName); os.IsNotExist(err) {
+		return "", fmt.Errorf("file %s does not exist", fileName)
+	}
+	buf, err := os.ReadFile(fileName)
+	if err != nil {
+		return "", fmt.Errorf("fail to read file %s: %w", fileName, err)
+	}
+	// fmt.Println(string(buf))
```
Comment (johnnyluo) [inline relay/state.go:38]: remove

File: demo/app/src/components/step-three.tsx [hunk 11]
```diff
@@ -22,15 +22,18 @@ export default function StepThree({ qrCodeString = '', session_id = '', uniqueSt
...
-    useEffect(() => {
-        const intervalId = setInterval(getDevices, 3000);
+    useEffect(  ()  =>  {
+        let intervalId: string | number | NodeJS.Timeout | undefined ;
+         postRoute(session_id, "VultiSignerApp").then(() => {
+              intervalId = setInterval(getDevices, 3000);
+         });
         return () => clearInterval(intervalId);
     }, [session_id]);
     // Since VultiSig does not support keygen, we ignored this peer id
```
Comment (johnnyluo) [inline demo/app/src/components/step-three.tsx:32]: can you please clean up this @0xkobe

File: dev_replace.go [hunk 12]
```diff
@@ -0,0 +1,13 @@
...
+
+package main
+
+import (
+	_ "github.com/vultisig/mobile-tss-lib"
+)
+
+func init() {
+	// This code is only included in the build when the 'dev' tag is specified.
+	_ = `go:generate go mod edit -replace=github.com/vultisig/mobile-tss-lib=../mobile-tss-lib`
```
Comment (johnnyluo) [inline dev_replace.go:12]: Can we please remove this?

File: internal/keygen/keygen.go [hunk 13]
```diff
@@ -1,25 +1,151 @@
...
 )
 
-func JoinKeyGeneration(kg *types.KeyGeneration) error {
-	relayServer := config.AppConfig.Relay.Server
+func JoinKeyGeneration(kg *types.KeyGeneration) (string, string, error) {
+	keyFolder := config.AppConfig.Server.VaultsFilePath
+	serverURL := config.AppConfig.Relay.Server
+
+	server := relay.NewServer(serverURL)
+	/*	if err := server.RegisterSession(kg.Session, kg.Key); err != nil {
```
Comment (johnnyluo) [inline internal/keygen/keygen.go:23]: remove it if we don't need it

File: internal/keygen/keygen.go [hunk 14]
```diff
@@ -1,25 +1,151 @@
...
+	}).Info("Session started")
 
-	_, err := coordinator.ExecuteKeyGeneration(coordinator.KeygenInput{
-		Server:    relayServer,
-		Key:       kg.Key,
-		Parties:   kg.Parties,
-		Session:   kg.Session,
-		ChainCode: kg.ChainCode,
+	if err != nil {
+		if err == context.DeadlineExceeded {
```
Comment (johnnyluo) [inline internal/keygen/keygen.go:37]: we are not doing anything special for `context.DeadlineExceeded` error , why not just return a wrapped error

File: relay/messenger.go [hunk 15]
```diff
@@ -0,0 +1,297 @@
...
+					}).Error("Failed to delete message, status code is not 200 OK")
+					continue
+				}
+			}
+		}
+	}
+}
+
+func encrypt(plainText, hexKey string) (string, error) {
+	key, _ := hex.DecodeString(hexKey)
```
Comment (johnnyluo) [inline relay/messenger.go:224]: don't swallow error , handle it appropriately

**Comment by Ehsan-saradar** [comment]
I suggest to fix e2e test or discard them for now to make sure all actions and pipelines are passed before merged

**Comment by 0xkobe** [comment]
> I suggest to fix e2e test or discard them for now to make sure all actions and pipelines are passed before merged

I agree to discard it for now.

File: test/e2e/keygen_test.go [hunk 16]
```diff
@@ -0,0 +1,129 @@
+package keygen_test
+
+// import (
+// 	"bytes"
```
Comment (johnnyluo) [inline test/e2e/keygen_test.go:4]: why the tests are commented out? if it is not working can we fix it? otherwise remove it

**Review by johnnyluo — APPROVED** [review]
LGTM
